File: Programmers/level2/formulaMax/formula_max.py
# 문제 출처 : https://programmers.co.kr/learn/courses/30/lessons/67257
# 참고 블로그 : https://eda-ai-lab.tistory.com/509
# 파이썬 정규식 레퍼런스 : https://python.flowdas.com/library/re.html
# 파이썬 permutation 레퍼런스 : https://python.flowdas.com/library/itertools.html?highlight=permutation#itertools.permutations

from itertools import permutations
import re


def solution(expression):
    answer = 0
    op = [x for x in ['*', '+', '-'] if x in expression]
    # print(op)
    op = [list(x) for x in permutations(op)]
    # print("op : ", op)
    exp = re.split(r'(\D)', expression)
    # print("exp : ", exp)
    stack = []
    for x in op:
        ex = exp[:]  # 새로운 배열로 복사
        # print("ex : ", ex)
        for y in x:
            # print(y)
            while y in ex:
                tmp = ex.index(y)
                # ex) 200 * 300
                ex[tmp-1] = str(eval(ex[tmp-1]+ex[tmp]+ex[tmp+1]))
                ex = ex[:tmp]+ex[tmp+2:]  # 연산 한 것은 제거
        # print("stack : ", stack)
        stack.append(ex[-1])
    print("result stack : ", stack)
    return max(abs(int(x)) for x in stack)

File: Programmers/level2/formulaMax/formulaMax.js
// https://programmers.co.kr/learn/courses/30/lessons/67257?language=javascript
// 참고 블로그 : https://yoon-dumbo.tistory.com/entry/프로그래머스-수식-최대화-javascript

export const solution = (expression) => {
    const permut = [
        ['*', '+', '-'],
        ['*', '-', '+'],
        ['+', '-', '*'],
        ['+', '*', '-'],
        ['-', '+', '*'],
        ['-', '*', '+'],
    ]
    let result = [];
    for(let op of permut){
        // ex) op = ['*', '+', '-']
        const temp = expression.split(/(\D)/)
        // temp = ["100", "-", "200", "*", "300", "-", "500", "+", "20"]
        for(let o of op){
            // ex ) o = '*'
            while(temp.includes(o)){
                // 현재 temp에 "*"가 있으니
                let idx = temp.indexOf(o);
                // '*'의 인덱스를 저장
                temp.splice(idx-1, 3, eval(temp.slice(idx-1, idx+2).join('')))
                // "200", "*", "300"를 slice로 자른 뒤 "200*300"으로 조인한 뒤 
                // eval로 문자열 계산을 하고 그 결과를 splice로 60000을 추가
                // temp = ["100", "-", "60000", "-", "500", "+", "20"]
            }
            // 그 다음인 op의 '+'를 위의 과정으로 반복
        }
        // 모든 op의 내용을 반복한 뒤 최종 결과 값을 result에 절대값으로 넣음
        result.push(Math.abs(temp[0]))
    }
    console.log(result)
    return Math.max(...result);
}
